modules/brc20_swap_index/lib/brc20_swap/decimal/decimal.go

```diff
@@ -0,0 +1,334 @@
+		for i := 0; i < n; i++ {
+			decimalPartStr += "0"
+		}
+		decimalPart, ok = new(big.Int).SetString(decimalPartStr, 10)
+		if !ok || decimalPart.Sign() < 0 {
+			return nil, fmt.Errorf("invalid decimal format: %s", parts[0])
+		}
+	}
+
+	value := new(big.Int).Mul(integerPart, precisionFactor[maxPrecision])
+	if value.Sign() < 0 {
+		value = value.Sub(value, decimalPart)
+	} else {
+		value = value.Add(value, decimalPart)
+	}
+
+	return &Decimal{Precition: uint(maxPrecision), Val: value}, nil
+}
+
+func MustNewDecimalFromString(s string, maxPrecision int) *Decimal {
+	parts := strings.Split(s, ".")
+	if len(parts) == 2 {
+		s = strings.TrimRight(s, "0")
+		s = strings.TrimRight(s, ".")
+	}
+	val, err := NewDecimalFromString(s, maxPrecision)
+	if err != nil {
+		panic(err)
+	}
+	return val
+}
+
+// String returns the string representation of a Decimal instance
+func (d *Decimal) String() string {
+	if d == nil {
+		return "0"
+	}
+	value := new(big.Int).Abs(d.Val)
+	quotient, remainder := new(big.Int).QuoRem(value, precisionFactor[d.Precition], new(big.Int))
+	sign := ""
+	if d.Val.Sign() < 0 {
+		sign = "-"
+	}
+	if remainder.Sign() == 0 {
+		return fmt.Sprintf("%s%s", sign, quotient.String())
+	}
+	decimalPart := fmt.Sprintf("%0*d", d.Precition, remainder)
+	decimalPart = strings.TrimRight(decimalPart, "0")
+	return fmt.Sprintf("%s%s.%s", sign, quotient.String(), decimalPart)
+}
+
+// Add adds two Decimal instances and returns a new Decimal instance
+func (d *Decimal) Add(other *Decimal) *Decimal {
+	if d == nil && other == nil {
+		return nil
+	}
+	if other == nil {
+		value := new(big.Int).Set(d.Val)
+		return &Decimal{Precition: d.Precition, Val: value}
+	}
+	if d == nil {
+		value := new(big.Int).Set(other.Val)
+		return &Decimal{Precition: other.Precition, Val: value}
+	}
+	if d.Precition != other.Precition {
+		panic("precition not match")
+	}
+	value := new(big.Int).Add(d.Val, other.Val)
+	return &Decimal{Precition: d.Precition, Val: value}
+}
+
+// Sub subtracts two Decimal instances and returns a new Decimal instance
+func (d *Decimal) Sub(other *Decimal) *Decimal {
+	if d == nil && other == nil {
+		return nil
+	}
+	if other == nil {
+		value := new(big.Int).Set(d.Val)
+		return &Decimal{Precition: d.Precition, Val: value}
+	}
+	if d == nil {
+		value := new(big.Int).Neg(other.Val)
+		return &Decimal{Precition: other.Precition, Val: value}
+	}
+	if d.Precition != other.Precition {
+		panic(fmt.Sprintf("precition not match, (%d != %d)", d.Precition, other.Precition))
+	}
+	value := new(big.Int).Sub(d.Val, other.Val)
+	return &Decimal{Precition: d.Precition, Val: value}
+}
+
+// Mul muls two Decimal instances and returns a new Decimal instance
+func (d *Decimal) Mul(other *Decimal) *Decimal {
+	if d == nil || other == nil {
+		return nil
+	}
+	value := new(big.Int).Mul(d.Val, other.Val)
+	// value := new(big.Int).Div(value0, precisionFactor[other.Precition])
+	return &Decimal{Precition: d.Precition, Val: value}
+}
```

Comment:
This seems wrong, as precision can't be derived just from the first one.

Seems like this needs more tests, similar to addition and subtraction.